# BAD_Mutations
Software package for predicting deleterious single nucleotide variants using sequence conservation across a phylogenetic sample.

## Overview
`BAD_Mutations` (**B**LAST-**A**ligned-**D**eleterious?) performs a likelihood ratio tset (LRT) for the prediction of deleterious single nucleotide variants, as described in [Chun and Fay (2009)](https://doi.org/10.1101/gr.092619.109). This approach uses locus-specific synonymous and nonsynonymous substitution rates, so it is only applicable to *protein coding* regions.

## Citation
If you use `BAD_Mutations` in your research, please cite the following:

> Kono, T.J., Lei, L., Shih, C.H., Hoffman, P.J., Morrell, P.L. and Fay, J.C., 2018. Comparative genomics approaches accurately predict deleterious variants in plants. G3: Genes, Genomes, Genetics, 8(10), pp.3321-3329. https://doi.org/10.1534/g3.118.200563

## Installation
Please see our [installation guide](https://github.com/MorrellLAB/BAD_Mutations/blob/master/Manual/Installation.md) for help with installation of `BAD_Mutations` and its dependencies.

## Usage
For usage instructions and example command lines, please see our [Wiki](https://github.com/MorrellLAB/BAD_Mutations/wiki) and [user manual](https://github.com/MorrellLAB/BAD_Mutations/wiki/Manual).
